libraries/Microsoft.Bot.Builder.Shared/AuthenticationHelper.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector.Authentication;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// Authentication helpers.
+    /// </summary>
+    public static class AuthenticationHelper
+    {
+        /// <summary>
+        /// Call context storage to propagate values throught the request.
+        /// </summary>
+        private static AsyncLocal<BotFrameworkAuthenticationContext> asyncLocal = new AsyncLocal<BotFrameworkAuthenticationContext>();
+
+        /// <summary>
+        /// Gets the request authentication context async.
+        /// </summary>
+        /// <param name="authHeader">The authentication header.</param>
+        /// <param name="httpClient">The HTTP client.</param>
+        /// <returns></returns>
+        public static async Task<BotFrameworkAuthenticationContext> GetRequestAuthenticationContextAsync(string authHeader, HttpClient httpClient)
+        {
+            ClaimsIdentity claimsIdentity;
+
+            if (string.IsNullOrEmpty(authHeader))
+            {
+                claimsIdentity = new ClaimsIdentity(new List<Claim>(), "anonymous");
+            }
+
+            bool usingEmulator = EmulatorValidation.IsTokenFromEmulator(authHeader);
```

Comment:
I know you didn't write this originally and I don't actually expect you to fix this since it's not directly related to what you're goals are with this PR, but, since I'm looking at it, I wanted to point out that this code path ends up parsing the raw `authHeader` multiple times into a `JwtToken` which is not only wasting CPU, but also creates a lot of extra garbage along the way that will need to be collected. Considering this is going to happen on every single call coming in through the adapter it's definitely something that should be optimized _at some point_.